source/mir/deser/msgpack.d

```diff
@@ -1,40 +1,865 @@
 /++
 $(H4 High level Msgpack deserialization API)
 
-This module requires msgpack-d package.
-
 Macros:
 IONREF = $(REF_ALTTEXT $(TT $2), $2, mir, ion, $1)$(NBSP)
 +/
 module mir.deser.msgpack;
+import mir.algebraic : Nullable;
+import mir.ser.msgpack : MessagePackFmt;
+import mir.ion.exception : IonErrorCode, ionException, ionErrorMsg;
+import mir.lob : Blob;
+
+struct MsgpackExtension
+{
+    Blob data;
+    ubyte type;
+}
+
+private static T unpackMsgPackVal(T)(scope ref const(ubyte)[] data)
+{
+    import std.traits : Unsigned;
+    alias UT = Unsigned!T;
+
+    if (data.length < UT.sizeof)
+    {
+        version (D_Exceptions)
+            throw IonErrorCode.unexpectedEndOfData.ionException;
+        else
+            assert(0, IonErrorCode.unexpectedEndOfData.ionErrorMsg);
+    }
+
+    UT ret = (cast(UT[1])cast(ubyte[UT.sizeof])data[0 .. UT.sizeof])[0];
+
+    version (LittleEndian)
+    {
+        import core.bitop : bswap, byteswap;
+        static if (T.sizeof >= 4) {
+            ret = bswap(ret);
+        } else static if (T.sizeof == 2) {
+            ret = byteswap(ret);
+        }
+    }
+
+    return cast(typeof(return))ret;
+}
+
+@safe @nogc pure
+private static void advance(scope ref const(ubyte)[] data, size_t newStart) {
+    if (data.length < newStart)
+    {
+        version (D_Exceptions)
+            throw IonErrorCode.unexpectedEndOfData.ionException;
+        else
+            assert(0, IonErrorCode.unexpectedEndOfData.ionErrorMsg);
+    }
+
+    data = data[newStart .. $];
+}
+
+private static void readMap(S)(ref S serializer, scope ref const(ubyte)[] data, size_t length)
+{
+    auto state = serializer.structBegin();
+    auto window = data[0 .. $];
+    foreach(i; 0 .. length)
+    {
+        if (window.length < 1)
+        {
+            version (D_Exceptions)
+                throw IonErrorCode.unexpectedEndOfData.ionException;
+            else
+                assert(0, IonErrorCode.unexpectedEndOfData.ionErrorMsg);
+        }
+
+        MessagePackFmt keyType = cast(MessagePackFmt)window[0];
+        window.advance(1);
+        uint keyLength = 0;
+        sw: switch (keyType)
```

Comment:
Also, timestamps and integer types need to be added. They should be converted to string of course.